### ISSUE TITLE
There is support for symlink / junctions created by mklink?

### ISSUE THREAD

**Issue by DragosFlorea** [jborean93/smbprotocol#87]
Hi,
I've seen there is support for symlink, I've tried to use it on linux and it seems that is working, but I have some issues with symlink and junctions created by mklink in windows with the following commands:
mklink /J D:\Share\LinkToFolder1 D:\Folder1
mklink /D D:\Share\LinkToFolder1 D:\Folder1

My Setup:
I have Folder Share and a folder named Folder1.
"Share" is used for smb share connection as root.
Inside Share I'm creating one junction or one symlink to Folder1

Both Share and Folder1 are on the same machine.

The error that I'm getting when I try to parse the files is the following:
![image](https://user-images.githubusercontent.com/19181854/112992908-4cc14a80-9171-11eb-8eca-17c3c2e9cf2b.png)

Thanks.

**Comment by jborean93** [comment]
So symbolic links and junctions point work to an extent, how they are handled depends on a few factors, I'm going to focus on Windows being the SMB Server and how it deals with them. Samba might be different depending on how the config on that side, IIRC they have the ability to have symbolic links act like junctions points do on Windows.

* Junction points are handled server side
  * Connecting to `\\server\share\junction\file.txt` is handled transparently by the server
* Symbolic links are handled client side
  * Connecting to `\\server\share\symlink\file.txt` or `\\server\share\symlink-file.txt` will result in a response with the status `STATUS_STOPPED_ON_SYMLINK` and the error containing the symbolic link details

When `smbprotocol` encounters a symbolic link  in the path it

* Parses the [reparse buffer data](https://docs.microsoft.com/en-us/openspecs/windows_protocols/ms-smb2/f15ae37d-a787-4bf2-9940-025a8c9c0022) contained in the error message
* Resolves the target path based on what it tried to open and what is in the `SubstituteName` (target path)
* If the resolved path does not start with `\\` (a local path) it raises the `SMBLinkRedirectionError` https://github.com/jborean93/smbprotocol/blob/9a141ec4e1d2a60819c9c560a9629ef90e5a932d/smbprotocol/exceptions.py#L811-L813
* If the resolved path server and/or share points to a different host/share than what was in the original connection it also fails
* If the resolved path points to the same server it will send another create request using the new UNC path specified

The reason why it validates the host/share is mostly due to the behaviour on the Windows clients and to simplify the codebase itself.

TLDR for this part: junction points should just work, they are not handled in any special way by the client. Symbolic links work as long as the target is the same hostname and share as where they were found.

The screenshot you have shared is less how `smbclient` handles symlinks when it tries to open a file but rather how it treats links encountered by `walk()` when walking the directory contents. Looking at the traceback I can see the following things happening:

* You called `smbclient.walk()` on some directory
* This essentially calls `smbclient.scandir()` for that path and any of it's directories to produce a dir entry object
  * This dir entry has a path which is `directory_path + '\' + entry_name`
* It has reached a dir entry that has the `FILE_ATTRIBUTE_REPARSE_POINT` attribute (symlinks and junctions are reparse points, amongst other types)
* It calls `smbclient.lstat(entry_path)` to get further information about that entry, `lstat` does not follow links so it is meant to get details on the link/object itself rather than it's target
* `smbclient.lstat()` essentially calls `smbclient.stat(path, follow_symlinks=False)`
* `smbclient.stat()` here tries to open the path with `FILE_OPEN_REPARSE_POINT`
* For some reason at this step it couldn't find an existing session for this host so it tried to reauthenticate itself
* Because no explicit credentials were specified when the walk dir entry called `lstat()` it is failing to use NTLM, Kerberos wasn't available for an environment reason as well

What is confusing is that each dir entry produced by `scandir()` should always have the same root as what the caller provided meaning the same connection/session/tree are reused without having to reauthenticate itself. At this point the symlink/junction hasn't been resolved at all so all the stuff I mentioned above the TLDR shouldn't apply right now in this case. Are you by chance using a `connection_cache` in the kwargs to the original `smbclient.walk()` call? That's the only reason that I can see that the existing session is not available.

**Comment by DragosFlorea** [comment]
Yeah you are right, I'm using `connection_cache `, because I'm using multiple threads in which I have other connections opened.
I have something like this
```
connection_cache = {}
try:
    register_session(share.Server, username=share.User, password=share.Password, encrypt=True, connection_cache = connection_cache )

    for root, dirs, fileNames in walk(sharePath,
                                          topdown=True,
                                          onerror=self.__onerror__,
                                          follow_symlinks=True,
                                          connection_cache = connection_cache ):
          ProcessFile()
catch:
...
finally
reset_connection_cache(connection_cache=connection_cache)
```
It seems that without `connection_cache`, "walk" is working fine through symlinks.
There is another option in which I can still use `connection_cache` in walk? because it was useful until now to have control of the opened connection

**Comment by jborean93** [comment]
Unfortunately no, the code in `walk/scandir` needs to be fixed up so that the connection cache is reused for subsequent calls. Right now it's a bug that it does not.
